### ISSUE TITLE
trouble with example/render_queue

### ISSUE THREAD

**Issue by cdelorme** [veandco/go-sdl2#203]
I am on linux, debian jessie specifically, and using the latest go (1.7).  I began playing with SDL2 and it works for me with the `runtime.LockOsThread()`, but obviously I would love to take advantage of multi-threading to process updates in parallel.

I looked into the examples, and the render_goroutines has a near-infinite length goroutine stacktrace, so I decided to post the crash I get from render_queue instead:

	$ go run render_queue.go 
	goroutine: C
	goroutine: B
	goroutine: A
	fatal error: unexpected signal during runtime execution
	[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x7fe564d2ea36]

	runtime stack:
	runtime.throw(0x4ca7e6, 0x2a)
		/home/cdelorme/.gvm/gos/go1.7/src/runtime/panic.go:566 +0x95
	runtime.sigpanic()
		/home/cdelorme/.gvm/gos/go1.7/src/runtime/sigpanic_unix.go:12 +0x2cc

	goroutine 5 [syscall, locked to thread]:
	runtime.cgocall(0x495970, 0xc420031f00, 0xc400000000)
		/home/cdelorme/.gvm/gos/go1.7/src/runtime/cgocall.go:131 +0x110 fp=0xc420031eb8 sp=0xc420031e78
	github.com/veandco/go-sdl2/sdl._Cfunc_SDL_RenderDrawPoint(0x18ef720, 0x12c00000096, 0x0)
		??:0 +0x4d fp=0xc420031f00 sp=0xc420031eb8
	github.com/veandco/go-sdl2/sdl.(*Renderer).DrawPoint(0x18ef720, 0x96, 0x12c, 0x0, 0x1)
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/sdl/render.go:399 +0x87 fp=0xc420031f48 sp=0xc420031f00
	main.run.func2()
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/examples/render_queue/render_queue.go:43 +0x57 fp=0xc420031f88 sp=0xc420031f48
	github.com/veandco/go-sdl2/sdl.processCalls()
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/sdl/sdl.go:38 +0xbe fp=0xc420031fc0 sp=0xc420031f88
	runtime.goexit()
		/home/cdelorme/.gvm/gos/go1.7/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc420031fc8 sp=0xc420031fc0
	created by github.com/veandco/go-sdl2/sdl.init.1
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/sdl/sdl.go:43 +0x35

	goroutine 1 [chan send]:
	main.run(0x0)
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/examples/render_queue/render_queue.go:66 +0x45f
	main.main()
		/home/cdelorme/.gvm/pkgsets/go1.7/global/src/github.com/veandco/go-sdl2/examples/render_queue/render_queue.go:127 +0x22

	goroutine 17 [syscall, locked to thread]:
	runtime.goexit()
		/home/cdelorme/.gvm/gos/go1.7/src/runtime/asm_amd64.s:2086 +0x1
	exit status 2

For all I know it could be a change to the latest go, but I figured I'd ask here in case it broke for other folks on earlier versions.  Has anyone gotten this to work in linux?

**Comment by cdelorme** [comment]
For amusement, I tried adding the lock os thread to the example file, and it still crashes.  Same with the render_goroutine example.  I will try switching go versions and give it another try.

**Comment by cdelorme** [comment]
Finished trying with go1.4.3, go1.5, and go1.6.2, all of which had the same crash.

If I didn't mention in the opening, the window appears for a fraction of a second before the crash occurs.

**Comment by cdelorme** [comment]
I could just be unskilled, and my backend knowledge of go is certainly lacking, but after a few hours of reading and trial/error I am confident that the CallQueue implementation is flawed.

The documentation for `runtime.LockOSThread()` states that it's context is the current goroutine and not the entire application:

> LockOSThread wires the **calling goroutine** to its current operating system thread

_I was operating under the false assumption that this meant new goroutines would be locked to the same thread and your application would no-longer scale across multiple cores._

It took several readings to make sense of [Russ Cox proposal](https://github.com/golang/go/wiki/LockOSThread) as well, as I had discarded the quote from the documentation in favor of my assumption, but these comments stuck out:

> Arrange that main.main runs on main thread

I assume this means that `func init()`, even when called from a dependent package, will be tied to `func main` in `package main`, which I assume starts on the "operating system thread".  Assuming that is the case then I would also assume that unless you use goroutines SDL should never fail, _which suggests that the goroutine for `package main` is transient (eg. scheduler relocates at runtime), or that the scheduler calls `init` directly but creates a goroutine for `main`, yielding potentially varied results unless you lock the thread from `init`._

> Main does not return

While implementation at the library was supplied this threw me for a loop as no application implementation made sense, until I realized the intention looks something like this:

	import "sdl"

	func main() {
		go func() {
			// initialize sdl
			// grab a window
			// grab a renderer
			// begin loop that makes calls to sdl.X
		}()
		sdl.Main()
	}

_The fact that `Main()` in this example does not return is because it is now ranging the mainfunc channel, which means until something like `sdl.Quit` closes the channel it locks the `main()` process._  As a result, all activity has to be sent to one or more goroutines **_before_** you call this, making it a considerable deviation from the current implementation (a breaking change for sure).

**_In conclusion, I believe that because sdl's `func init()` runs `go processCalls()` the extra goroutine leads to an invalid implementation, as the new goroutine is in no way guaranteed to be associated with the same thread as `main.main`._**  What I am not sure of is whether "current operating system thread" means the thread that the goroutine was scheduled to run on or if it should mean the same thread as `main()`, and whether merely having another goroutine touching sdl (even from the same thread) creates new potential conflicts that cause errors (perhaps at the scheduler?).

On another note, I did create an example that uses goroutines to pass functions that refer to a renderer to the main thread to execute, _effectively the inverse of Russ Cox proposal_:

	package main

	import (
		"fmt"
		"runtime"
		"time"

		"github.com/veandco/go-sdl2/sdl"
	)

	var running bool
	var window *sdl.Window
	var renderer *sdl.Renderer

	func init() {
		runtime.LockOSThread()
	}

	func main() {
		var err error
		if err = sdl.Init(sdl.INIT_VIDEO); err != nil {
			fmt.Println(err.Error())
			return
		}

		if window, err = sdl.CreateWindow("Rendering with goroutines", sdl.WINDOWPOS_CENTERED, sdl.WINDOWPOS_CENTERED, 800, 600, sdl.WINDOW_SHOWN); err != nil {
			fmt.Println(err.Error())
			return
		}

		if renderer, err = sdl.CreateRenderer(window, -1, sdl.RENDERER_ACCELERATED); err != nil {
			fmt.Println(err.Error())
			defer window.Destroy()
			return
		}

		// start with a black canvas
		renderer.Clear()
		renderer.Present()

		// setup async stack to queue input checks and multiple rendering loops
		operations := make(chan func())
		running = true
		go input(operations)
		go render(operations, 1)
		go render(operations, 2)
		go render(operations, 3)

		for f := range operations {
			f()
		}

		sdl.Quit()
	}

	func render(c chan<- func(), num int) {

		// define a list of functions and an index
		var list []func()
		index := num

		// add all the contents that can be drawn here, with a reference to the package global renderer
		list = append(list, func() {
			renderer.Clear()
			renderer.SetDrawColor(255, 255, 255, 255)
			renderer.DrawPoint(150, 300)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			renderer.SetDrawColor(0, 0, 255, 255)
			renderer.DrawLine(0, 0, 200, 200)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			renderer.SetDrawColor(255, 255, 255, 255)
			renderer.DrawPoint(150, 300)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			points := []sdl.Point{{0, 0}, {100, 300}, {100, 300}, {200, 0}}
			renderer.SetDrawColor(255, 255, 0, 255)
			renderer.DrawLines(points)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			rect := sdl.Rect{300, 0, 200, 200}
			renderer.SetDrawColor(255, 0, 0, 255)
			renderer.DrawRect(&rect)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			rects := []sdl.Rect{{400, 400, 100, 100}, {550, 350, 200, 200}}
			renderer.SetDrawColor(0, 255, 255, 255)
			renderer.DrawRects(rects)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			rect := sdl.Rect{250, 250, 200, 200}
			renderer.SetDrawColor(0, 255, 0, 255)
			renderer.FillRect(&rect)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		list = append(list, func() {
			renderer.Clear()
			rects := []sdl.Rect{{500, 300, 100, 100}, {200, 300, 200, 200}}
			renderer.SetDrawColor(255, 0, 255, 255)
			renderer.FillRects(rects)
			renderer.Present()
			fmt.Printf("Change from channel %d\n", num)
		})

		for running {
			time.Sleep(time.Second * time.Duration(num))
			c <- list[index]
			index = (index + 1) % len(list)
		}
		fmt.Printf("Channel %d closing...\n", num)
	}

	func input(c chan func()) {
		for running {
			c <- func() {
				var event sdl.Event
				for event = sdl.PollEvent(); event != nil; event = sdl.PollEvent() {
					switch event.(type) {
					case *sdl.QuitEvent:
						running = false
					case *sdl.KeyDownEvent:
						if event.(*sdl.KeyDownEvent).Keysym.Sym == sdl.K_ESCAPE {
							running = false
						}
					}
				}
			}
		}
		fmt.Println("Input captured exit...")
		close(c)
	}

**The main problem with this solution is it lacks a meaningful (or simple) way to abstract return values,** which is not something I would care to try.

Since an implementation at the library involves a significant change, I am curious whether one of these would be preferable:

- give developers the choice by adding prefix or suffix functions to wrap thread-locked behavior
- create a separate wrapper package that provides this behavior without adding to the current library

I would be happy to contribute to either approach, but I can't promise the time necessary to take ownership of maintaining it.

**_Also, even if the sdl library added `runtime.LockOSThread()` to its `init`, it should not effect the rest of a developers application and should be added by default to correct some of the more basic problems with transient/startup behavior._**  My guess is application state could be updated from a goroutine with just that, so long as all calls to sdl are only run from the main thread.

**Comment by veeableful** [comment]
Hi @cdelorme, I've updated the `render_goroutines` example. Could you check if the crash is still happening? I've wrapped all the SDL2 functions under the CallQueue so it shouldn't crash anymore.

**Comment by cdelorme** [comment]
@veeableful Your patch worked, the code now displays a series of white outlined boxes sliding across a black background.

I'm still not sure I follow how this works, but does this mean that so long as only one goroutine/thread uses sdl functions it doesn't matter which one calls `runtime.LockOSThread()`?

**Comment by veeableful** [comment]
@cdelorme According to SDL2's [article](https://wiki.libsdl.org/CategoryThread), as long as SDL2's thread-sensitive functions are called from the same OS thread (which is the goroutine that calls `runtime.LockOSThread`), it should be safe.

How it works in the library is that we pass functions to that "main" goroutine which then becomes the one who calls the SDL2's thread-sensitive functions. It is also synchronous as it executes the function as soon as it receives from the user.

Although I do have a thought that the crash you experienced may not be caused by that problem at all. It might be caused by the 800 x 600 goroutines that were spawned to fill the screen with random colors (which doesn't crash on my chromebook). I don't have the Go technical knowledge to confirm it though..

I'm also thinking that it would be much nicer to make all SDL2 thread-sensitive functions use the `CallQueue` so the user doesn't have to know about it at all. I might start a different branch (with different build tag) which implements that and see how the performance is affected.

**Comment by cdelorme** [comment]
@veeableful Thank you for the detailed explanation; I figured there was a particular set of rules in SDL around thread association but I couldn't find the document on their wiki that described that, so the link is appreciated.

I did notice that every call to SDL was being routed through the queue stack which was locked to a thread.  I thought it was the main thread (eg. first thread the program starts with, and thus tied to `func main()`) that had to be in charge of making the calls, not just "the same thread has to make all calls in to sdl", which is how it appears to be working with the CallQueue.

I was testing from a rather beefy linux desktop (16GB DDR31600, Core i7 4790, nVidia GTX 970) so if you got the previous version working on a laptop I doubt it's the number of goroutines.  I may checkout the old version on my macbook and give it a shot there to see if it's any different.

If the performance hit is negligable, I think it would be wonderful to create that abstraction so that developers don't have to worry which goroutine calls in to sdl.  If you'd like help coding/testing let me know.

I'll leave this ticket open since the `render_queue` example is still broken for me.

**Comment by cdelorme** [comment]
I stand corrected, I gave the previous version a try using my mac and it worked.  It looks like a big static picture that appeared suddenly so you couldn't really tell goroutines were used to draw the screen.

**Comment by veeableful** [comment]
I've updated the `render_queue` example. Could you check if it works?

and of course I'd like help! If you could fork this repository, we could start working on the SDL2 thread-sensitive functions under the CallQueue. Once it's tested and we're happy with it, I'd merge it!

**Comment by cdelorme** [comment]
I tested the old and new on my macbook as well as the linux desktop, and the old one broke on both, the new one works on both.  I will close this ticket with this comment as the issue is resolved.  Thank you again; I think I have a much better understanding of how this all ties together now.

I created a fork and added you as a collaborator if that's agreeable.  I'll try to review the list of sdl functions related to rendering, window, and event management and get some of them wrapped for testing this evening.
